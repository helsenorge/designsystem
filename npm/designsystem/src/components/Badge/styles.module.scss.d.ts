export type Styles = {
  badge: string;
  'badge--blueberry': string;
  'badge--cherry': string;
  'badge--neutral': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
