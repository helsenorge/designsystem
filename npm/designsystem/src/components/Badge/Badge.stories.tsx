import React from 'react';

import { Meta, StoryObj } from '@storybook/react';

import Badge from './Badge';
import Docs from '../../docs';
import Title from '../Title';

const meta = {
  title: '@helsenorge/designsystem-react/Components/Badge',
  component: Badge,
  parameters: {
    docs: {
      page: (): React.JSX.Element => <Docs component={Badge} />,
      description: {
        component:
          'Badge [Markør] lar innbygger oppfatte at det er et antall nye elementer som har kommet til i et område siden sist gang innbygger besøkte området.',
      },
    },
  },
  args: {
    children: 'Tekst',
    color: 'blueberry',
    type: 'string',
    notificationVariant: 'info',
  },
  argTypes: {
    children: {
      control: 'text',
    },
    color: {
      control: 'select',
      options: ['blueberry', 'cherry', 'neutral'],
    },
    type: {
      control: 'select',
      options: ['string', 'notification'],
    },
    notificationVariant: {
      control: 'select',
      options: ['success', 'warning', 'error', 'info'],
    },
  },
} satisfies Meta<typeof Badge>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: args => <Badge {...args} />,
};

export const Variants: Story = {
  render: args => (
    <>
      <Title margin={2} htmlMarkup={'h3'} appearance={'title3'}>
        {'Variants'}
      </Title>
      <br />
      <Badge color="blueberry">{args.children}</Badge>
      <br />
      <Badge color="cherry">{args.children}</Badge>
      <br />
      <Badge color="neutral">{args.children}</Badge>
    </>
  ),
};

export const Notification: Story = {
  render: () => (
    <>
      <Title margin={2} htmlMarkup={'h3'} appearance={'title3'}>
        {'Notification variants'}
      </Title>
      <br />
      <Badge type="notification" notificationVariant="info" />
      <br />
      <Badge type="notification" notificationVariant="warn" />
      <br />
      <Badge type="notification" notificationVariant="error" />
      <br />
      <Badge type="notification" notificationVariant="success" />
    </>
  ),
};
