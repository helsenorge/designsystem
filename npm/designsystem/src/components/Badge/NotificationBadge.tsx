import React from 'react';

import { IconSize } from '../../constants';
import { palette } from '../../theme/palette';
import Icon from '../Icon';
import CheckFill from '../Icons/CheckFill';
import ErrorSignFill from '../Icons/ErrorSignFill';
import InfoSignFill from '../Icons/InfoSignFill';
import TriangleX from '../Icons/TriangleX';

export type BadgeNotificationVariant = 'info' | 'warn' | 'error' | 'success';

export interface NotificationBadgeProps {
  /** Changes the type of the badge */
  variant: BadgeNotificationVariant;
  /** Sets the size of the badge */
  size: IconSize;
  /** Adds custom classes to the element */
  className?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
  /** Sets hover state */
  isHovered?: boolean;
}

const NotificationBadge = (props: NotificationBadgeProps): React.JSX.Element => {
  const { variant, size, className = '', testId, isHovered = false } = props;
  switch (variant) {
    case 'info':
      return (
        <Icon
          svgIcon={InfoSignFill}
          color={palette.blueberry700}
          hoverColor={palette.blueberry700}
          size={size}
          className={className}
          testId={testId}
          isHovered={isHovered}
        />
      );
    case 'warn':
      return (
        <Icon
          svgIcon={ErrorSignFill}
          color={palette.banana700}
          hoverColor={palette.banana700}
          size={size}
          className={className}
          testId={testId}
          isHovered={isHovered}
        />
      );
    case 'error':
      return (
        <Icon
          svgIcon={TriangleX}
          color={palette.cherry700}
          hoverColor={palette.cherry700}
          size={size}
          className={className}
          testId={testId}
          isHovered={isHovered}
        />
      );
    case 'success':
      return (
        <Icon
          svgIcon={CheckFill}
          color={palette.kiwi900}
          hoverColor={palette.kiwi900}
          size={size}
          className={className}
          testId={testId}
          isHovered={isHovered}
        />
      );
  }
};

export default NotificationBadge;
