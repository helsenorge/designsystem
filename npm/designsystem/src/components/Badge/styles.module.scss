@import '../../scss/palette';

.badge {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  font-size: 0.875rem;
  font-weight: 600;
  min-width: 1.375rem;
  height: 1.375rem;
  border-radius: 1rem;
  padding: 0 0.375rem;
  color: $white;
  background-color: $black;

  &--white {
    color: $black;
    background-color: $white;
  }

  &--blueberry {
    background-color: $blueberry500;
  }

  &--banana {
    background-color: $banana500;
  }

  &--cherry {
    background-color: $cherry500;
  }

  &--kiwi {
    background-color: $kiwi500;
  }

  &--neutral {
    background-color: $neutral500;
  }

  &--plum {
    background-color: $plum500;
  }
}
