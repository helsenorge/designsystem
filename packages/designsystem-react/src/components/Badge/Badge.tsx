import React from 'react';
import {PaletteNames} from '../../theme/palette';
import StyledBadge from './Badge.styled';

export type BadgeColors = PaletteNames;
export type BadgeChildren = string | number;

interface BadgeProps {
  children: BadgeChildren;
  className?: string;
  color?: BadgeColors;
}

const Badge = React.forwardRef((props: BadgeProps, ref: any) => {
  const {children, className = '', color = 'black'} = props;
  const oversized = children.toString().length > 2;
  return (
    <StyledBadge oversized={oversized} className={className} color={color} ref={ref}>
      {children}
    </StyledBadge>
  );
});

export default Badge;
