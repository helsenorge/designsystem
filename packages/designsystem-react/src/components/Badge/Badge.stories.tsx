import React from 'react';
import {withA11y} from '@storybook/addon-a11y';
import {withKnobs, text, select} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import Badge from './Badge';
import {allPaletteNames} from '../../../.storybook/knobs';

const stories = storiesOf('Badge', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withA11y);

stories.add('Default', () => <Badge color={select('Color', allPaletteNames, 'blueberry')}>{text('Text', '1')}</Badge>);
