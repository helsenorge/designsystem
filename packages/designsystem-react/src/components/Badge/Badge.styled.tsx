import {BadgeColors} from './Badge';
import styled from 'styled-components';
import {getColor} from '../../theme/currys';

interface StyledBadgeProps {
  color: BadgeColors;
  oversized: boolean;
}

const StyledBadge = styled('span')<StyledBadgeProps>`
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 0 ${props => (props.oversized ? '0.25rem' : 0)};
  font-size: 0.875rem;
  font-weight: 600;
  color: ${props => (props.color === 'white' ? 'black' : 'white')};
  min-width: 1.375rem;
  height: 1.375rem;
  border-radius: 1rem;
  background-color: ${props => (props.color ? getColor(props.color, 500) : getColor('black'))};
`;

export default StyledBadge;
