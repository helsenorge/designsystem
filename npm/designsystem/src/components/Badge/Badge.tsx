import React from 'react';

import classNames from 'classnames';

import NotificationBadge, { BadgeNotificationVariant } from './NotificationBadge';
import { AnalyticsId, IconSize } from '../../constants';
import { PaletteNames } from '../../theme/palette';

import badgeStyles from './styles.module.scss';

export type BadgeColors = Extract<PaletteNames, 'blueberry' | 'cherry' | 'neutral'>;
export type BadgeChildren = string | number;
export type BadgeVariant = 'string' | 'notification';

export interface BadgeProps {
  /** Sets the content of the badge. */
  children?: BadgeChildren;
  /** Adds custom classes to the element. */
  className?: string;
  /** Changes the badge background color. */
  color?: BadgeColors;
  /** Sets the data-testid attribute. */
  testId?: string;
  /** Changes the type of the badge. Default: string */
  type?: BadgeVariant;
  /** Type of notification badge. Only works if type is 'notification'. */
  notificationVariant?: BadgeNotificationVariant;
}

export type BadgeType = React.ForwardRefExoticComponent<BadgeProps & React.RefAttributes<HTMLElement>>;
const Badge: BadgeType = React.forwardRef(function BadgeForwardedRef(props: BadgeProps, ref: React.ForwardedRef<HTMLElement>) {
  const { children, className = '', color = 'blueberry', testId, type = 'string', notificationVariant = 'info' } = props;
  const badgeClasses = classNames(
    badgeStyles.badge,
    {
      [badgeStyles['badge--blueberry']]: color === 'blueberry',
      [badgeStyles['badge--cherry']]: color === 'cherry',
      [badgeStyles['badge--neutral']]: color === 'neutral',
    },
    className
  );

  if (type === 'notification' && notificationVariant !== undefined) {
    return (
      <NotificationBadge
        variant={notificationVariant}
        size={IconSize.XSmall}
        className={className}
        testId={testId}
        data-analyticsid={AnalyticsId.Badge}
      />
    );
  }

  return (
    <span className={badgeClasses} ref={ref} data-testid={testId} data-analyticsid={AnalyticsId.Badge}>
      {children}
    </span>
  );
});

export default Badge;
