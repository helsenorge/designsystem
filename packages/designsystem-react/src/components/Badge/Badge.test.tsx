import React from 'react';
import {render} from '@testing-library/react';
import Badge from './Badge';
// import 'jest-styled-components';
// import {palette} from '../../theme';

const badgeNumber = 123;

test('displays text', (): void => {
  const {container, getByText} = render(<Badge>{badgeNumber}</Badge>);
  expect(getByText(badgeNumber.toString())).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

/* 
  TODO: Waiting for issues regarding babel and types for this to be implemented
  Seems to be version issues because of v5 release.
*/
// test('has the right color', (): void => {
//   const {container} = render(<Badge color="blood200">{badgeNumber}</Badge>);
//   expect(container).toMatchSnapshot();
// });
