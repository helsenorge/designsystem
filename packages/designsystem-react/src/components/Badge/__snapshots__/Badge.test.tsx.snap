// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays text 1`] = `
<div>
  <span
    class="sc-AxjAm drUKFi"
    color="black"
  >
    123
  </span>
</div>
`;
